File: app/src/main/java/com/itba/runningMate/mainpage/fragments/running/services/location/TrackingServiceBinder.java
package com.itba.runningMate.mainpage.fragments.running.services.location;

import android.os.Binder;

import com.itba.runningMate.mainpage.fragments.running.model.Route;
import com.itba.runningMate.utils.run.RunMetrics;

public class TrackingServiceBinder extends Binder implements Tracker {

    private final TrackingService trackingService;

    public TrackingServiceBinder(final TrackingService trackingService) {
        this.trackingService = trackingService;
    }

    public Route queryRoute() {
        return new Route().addLatLngsToRoute(trackingService.getTrackedLocations());
    }

    public boolean isTracking() {
        return trackingService.isTracking();
    }

    public void startTracking() {
        trackingService.startTracking();
    }

    public void stopTracking() {
        trackingService.stopTracking();
    }

    public void setOnTrackingUpdateListener(OnTrackingUpdateListener listener) {
        trackingService.setOnTrackingUpdateListener(listener);
    }

    @Override
    public void setOnTrackingLocationUpdateListener(OnTrackingLocationUpdateListener listener) {
        trackingService.setOnTrackingLocationUpdateListener(listener);
    }

    @Override
    public void setTrackingMetricsUpdateListener(OnTrackingMetricsUpdateListener listener) {
        trackingService.setOnTrackingMetricsUpdateListener(listener);
    }

    @Override
    public void removeTrackingUpdateListener(OnTrackingUpdateListener listener) {
        trackingService.removeTrackingUpdateListener(listener);
    }

    @Override
    public void removeTrackingLocationUpdateListener(OnTrackingLocationUpdateListener listener) {
        trackingService.removeTrackingLocationUpdateListener(listener);
    }

    @Override
    public void removeTrackingMetricsUpdateListener(OnTrackingMetricsUpdateListener listener) {
        trackingService.removeTrackingMetricsUpdateListener(listener);
    }

    @Override
    public long queryStartTime() {
        return trackingService.getStartTimeMillis();
    }

    @Override
    public float queryDistance() {
        return trackingService.getElapsedDistance();
    }

    @Override
    public long queryElapsedTime() {
        return trackingService.getElapsedMillis();
    }

    @Override
    public long queryPace() {
        return trackingService.getPace();
    }

    @Override
    public float queryVelocity() {
        return RunMetrics.calculateVelocity(trackingService.getElapsedDistance(), trackingService.getElapsedMillis());
    }
}

File: app/src/main/java/com/itba/runningMate/services/location/TrackingServiceBinder.kt
package com.itba.runningMate.services.location

import android.os.Binder
import com.itba.runningMate.domain.Route
import com.itba.runningMate.services.location.listeners.OnTrackingLocationUpdateListener
import com.itba.runningMate.services.location.listeners.OnTrackingMetricsUpdateListener
import com.itba.runningMate.services.location.listeners.OnTrackingUpdateListener
import com.itba.runningMate.utils.run.RunMetrics

class TrackingServiceBinder(private val trackingService: TrackingService) : Binder(), Tracker {

    override fun queryRoute(): Route {
        return Route().addLatLngsToRoute(trackingService.getTrackedLocations())
    }

    override fun isTracking(): Boolean {
        return trackingService.isTracking
    }

    override fun startTracking() {
        trackingService.startTracking()
    }

    override fun stopTracking() {
        trackingService.stopTracking()
    }

    override fun newLap() {
        trackingService.newLap()
    }

    override fun resumeTracking() {
        trackingService.resumeTracking()
    }

    override fun setOnTrackingUpdateListener(listener: OnTrackingUpdateListener) {
        trackingService.setOnTrackingUpdateListener(listener)
    }

    override fun setOnTrackingLocationUpdateListener(listener: OnTrackingLocationUpdateListener) {
        trackingService.setOnTrackingLocationUpdateListener(listener)
    }

    override fun setTrackingMetricsUpdateListener(listener: OnTrackingMetricsUpdateListener) {
        trackingService.setOnTrackingMetricsUpdateListener(listener)
    }

    override fun removeTrackingUpdateListener(listener: OnTrackingUpdateListener) {
        trackingService.removeTrackingUpdateListener(listener)
    }

    override fun removeTrackingLocationUpdateListener(listener: OnTrackingLocationUpdateListener) {
        trackingService.removeTrackingLocationUpdateListener(listener)
    }

    override fun removeTrackingMetricsUpdateListener(listener: OnTrackingMetricsUpdateListener) {
        trackingService.removeTrackingMetricsUpdateListener(listener)
    }

    override fun queryStartTime(): Long {
        return trackingService.startTimeMillis
    }

    override fun queryEndTime(): Long {
        return trackingService.endTimeMillis
    }

    override fun queryDistance(): Float {
        return trackingService.elapsedDistance
    }

    override fun queryElapsedTime(): Long {
        return trackingService.runningMillis
    }

    override fun queryPace(): Long {
        return trackingService.pace
    }

    override fun queryVelocity(): Float {
        return RunMetrics.calculateVelocity(
            trackingService.elapsedDistance,
            trackingService.runningMillis
        )
    }
}